gateway-server/server.py

```diff
@@ -0,0 +1,91 @@
+from flask import Flask, request, jsonify, render_template
+import requests
+# TODO: from kazoo.client import KazooClient
+from pymongo import MongoClient
+import random
+
+app = Flask(__name__)
+
+# TODO: Add Zookeeper functionality
+
+mongo_uri = '<hidden-for-pull-request>'
+client = MongoClient(mongo_uri)
+db = client['nabu']
+
+# List of IPFS node addresses
+IPFS_URL = [
+    'http://localhost:5001',
+    # TODO: Add more IPFS node addresses
+]
+
+# IPFS routes
+PUT_ROUTE = '/api/v0/block/put'
+GET_ROUTE = '/api/v0/block/get'
+
+# Web UI
+@app.route('/')
+def index():
+    return render_template('index.html')
+
+# Forward GET request to IPFS peer to retrieve all content
+@app.route('/ipfs', methods=['GET'])
+def get_ipfs_content():
+    # Retrieve CIDs from database
+    cid_records = db['cid'].find()
+    cid_list = [record['cid'] for record in cid_records]
+    # Sample CIDs for tracing
+    nsamples = max(1, len(cid_list) // 10)
+    traced = random.sample(range(len(cid_list)), nsamples)
+
+    responses = []
+
+    for i, cid in enumerate(cid_list):
+        responses.append(send_single_get_request(cid, i in traced))
```

Comment:
This is really serialized. Can we do it in parallel so that we can stress the backend with several parallel requests? Okay to add a TODO and make this change in next PR.

Reply:
+1.

I think this could be done in parallel or concurrently with a threadpool executor: https://docs.python.org/3/library/concurrent.futures.html